server/routes/timetableItemBuilder.ts

```diff
@@ -0,0 +1,184 @@
+import { format, parseISO } from 'date-fns'
+import { SessionSchedule } from '../data/orchestrationApiTypes'
+import { prisonerTimePretty } from '../utils/utils'
+
+export type TimetableItem = {
+  time: string
+  type: 'Open' | 'Closed'
+  capacity: string
+  attendees: string
+  frequency: string
+  endDate: string
+}
+// Builds timetable rows, using all session schedules for the selected date
+export default ({
+  schedules,
+  selectedDate,
+}: {
+  schedules: SessionSchedule[]
+  selectedDate: string
+}): TimetableItem[] => {
+  const dateFormat = 'd MMMM yyyy'
+  const timetableItems: TimetableItem[] = []
+  schedules.forEach(schedule => {
+    const time = `${prisonerTimePretty(`${selectedDate}T${schedule.sessionTimeSlot.startTime}`)} to ${prisonerTimePretty(`${selectedDate}T${schedule.sessionTimeSlot.endTime}`)}`
+
+    const endDate = schedule.sessionDateRange.validToDate
+      ? format(parseISO(schedule.sessionDateRange.validToDate), dateFormat)
+      : 'Not entered'
+
+    let frequency = ''
+    if (
+      schedule.sessionDateRange.validToDate &&
+      schedule.sessionDateRange.validFromDate === schedule.sessionDateRange.validToDate
+    ) {
+      frequency = 'One off'
+    } else {
+      frequency = schedule.weeklyFrequency === 1 ? 'Every week' : `Every ${schedule.weeklyFrequency} weeks`
+    }
+
+    if (schedule.capacity.open !== 0) {
+      timetableItems.push({
+        time,
+        type: 'Open',
+        capacity: `${schedule.capacity.open} tables`,
+        attendees: buildAttendeesText(schedule),
+        frequency,
+        endDate,
+      })
+    }
+
+    if (schedule.capacity.closed !== 0) {
+      timetableItems.push({
+        time,
+        type: 'Closed',
+        capacity: `${schedule.capacity.closed} tables`,
+        attendees: buildAttendeesText(schedule),
+        frequency,
+        endDate,
+      })
+    }
+  })
+
+  return timetableItems
+}
+
+// Takes all group names for particular type, and joins together
+const mergeGroupNames = (groupNames: string[]): string => {
+  let nameString = ''
+  groupNames.forEach((name, index) => {
+    if (index === 0) {
+      nameString = name
+    } else if (index === groupNames.length - 1) {
+      nameString = `${nameString} and ${name}`
+    } else {
+      nameString = `${nameString}, ${name}`
```

Comment:
You can do all this as a join, then replace the last `,` with `and` ? Something like that, assuming group names don't have commas?

Reply:
Or if not 100% sure, join on `ZZZ` (mad string that won't be there), the replace the last `ZZZ` with `and` then replace `ZZZ` with `,`